module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php

```diff
@@ -363,33 +365,30 @@ protected function getItemStatus(
      *
      * @param array  $record            Information on items linked to a single
      *                                  bib record
-     * @param array  $messages          Custom status HTML
-     *                                  (keys = available/unavailable/uncertain/unknown)
      * @param string $callnumberSetting The callnumber mode setting used for
      *                                  pickValue()
      *
      * @return array                    Summarized availability information
      */
-    protected function getItemStatusGroup($record, $messages, $callnumberSetting)
+    protected function getItemStatusGroup($record, $callnumberSetting)
     {
+        // Get combined availability
+        $combinedInfo = $this->availabilityStatusManager->combine($record);
+        $combinedAvailability = $combinedInfo['availability'];
+
         // Summarize call number, location and availability info across all items:
         $locations = [];
-        $useUnknownStatus = false;
-        $available = 'false';
         foreach ($record as $info) {
+            $availabilityStatus = $info['availability'];
             // Find an available copy
-            if ($info['availability']) {
-                $availStr = $this->availabilityToString($info['availability']);
-                if ('true' !== $available) {
-                    $available = $availStr;
-                }
+            if ($availabilityStatus->isAvailable()) {
                 if ('true' !== ($locations[$info['location']]['available'] ?? null)) {
-                    $locations[$info['location']]['available'] = $availStr;
+                    $locations[$info['location']]['available'] = $availabilityStatus->getStatusDescription();
                 }
             }
             // Check for a use_unknown_message flag
-            if ($info['use_unknown_message'] ?? false) {
-                $useUnknownStatus = true;
+            if ($availabilityStatus->useUnknownMessage()) {
```

Comment:
Should this logic be moved inside the availability status manager's `combine` method? I realize that it is probably more efficient to do it here to avoid an extra loop through the statuses, but it feels strange doing it outside of the class that encapsulates status-related logic.

Reply:
When I changed this I was wondering how much sense this logic makes anyway. But I didn't change it in case it breaks something.

We set 'use_unknown_message' to true for the status with highest priority if any status has this. This means that the 'unkown' status has the highest priority when selecting the ajax status template. I don't know if there are other implications. But shouldn't an 'available' status have higher priority than 'unknown'?  Should we move this into `getPriority` in the `AvailabiltiyStatus` like
```
switch ($this->availability) {
            if ($this->useUnknownMessage) {
                return 2;
            }
            case AvailabilityStatus::STATUS_UNAVAILABLE:
                return 0;
            case AvailabilityStatus::STATUS_UNCERTAIN:
                return 1;
            default:
                return 3;
        }
```

Reply:
I think the problem here is that the use_unknown_message parameter was added to the ILS driver spec when availability was still a binary setting. The intention is that if it is set, it takes priority over the availability setting, and regardless of whether that value is set to true or false, it is not meaningful. Now that we have this richer mechanism for representing statuses, it may make the most sense to deprecate use_unknown_message entirely and instead return a more appropriate value in the availability key.

I believe its original use case was for the NoILS driver's oflfine mode, as a way to indicate status information was unavailable while the primary ILS driver was disabled due to an outage. It looks like a couple of other drivers (Voyager and XCNCIP) have made use of it in the meantime. I don't think much thought was given to a scenario where some items have unknown status and others have known availability, because the original use case was for an all-or-nothing scenario.

Bottom line: we should simplify this, and it might not even take a lot of work to do it. But some of that work may be out of scope for this PR, so taking a simple incremental step forward is probably fine.

Reply:
Theoretically, it would be possible to have some custom logic where there is a difference between an available and unavailable status that have both use_unknown_message=true. That's why I did not change this in the first place. But probably no one did this. And I was wondering what the difference between an STATUS_UNKNOWN and the STATUS_UNCERTAIN would be. 

For now, I moved the logic into `getPriority`.

Reply:
I think the difference is this:

STATUS_UNKNOWN = the system that provides status information is offline, so we cannot look up status information

STATUS_UNCERTAIN = the status information system is online, but it is unable to provide enough information to determine availability

These names do not clearly communicate this difference, though. Maybe STATUS_SYSTEM_OFFLINE would be a better name than STATUS_UNKNOWN. I think the main question is whether we want to try to fix this now, or accept incremental progress and address it separately.

@EreMaijala, any additional thoughts or corrections?

Reply:
@demiankatz Sort of like that, but not necessarily. :) As I see it, the main difference between STATUS_UNKNOWN and STATUS_UNCERTAIN is that with STATUS_UNCERTAIN we have an item with some status information but can't really determine if it's available or not, so we'll provide a yellow status indicator, whereas with STATUS_UNKNOWN we could have e.g. holdings-level information but no item-level availability information.

Since STATUS_UNCERTAIN is relatively new, it's possible that STATUS_UNKNOWN has also been used in cases where STATUS_UNCERTAIN would be more appropriate.

Reply:
True, but maybe that's all the more reason to straighten it out, now that it's possible to differentiate between these cases, and while we're revising the code anyway. This seems like something that could be handled by a note in the changelog (e.g. we could deprecate the unknown status key in the ILS driver spec and recommend using an availability value instead). I'll leave it up to @ThoWagen whether it's better to merge this first and do that separately, or try to integrate all the work.

Reply:
I replaced 'useUnknownStatus' by `STATUS_UNKNOWN` in `AvailabilityStatus` . I did not touch the drivers and check_item_statuses.js though. I think that should be done in another PR.

Reply:
Yes, I agree, as long as the higher-level code translates things in a reasonable fashion, that is good enough for this PR. Cleaning things up and making them consistent in lower-level code makes sense as a follow-up action.

I'll try to give all of this a more thorough review next week when my mind is a bit fresher. :-)